src/h3lib/lib/h3Index.c

```diff
@@ -78,49 +78,108 @@ void H3_EXPORT(h3ToString)(H3Index h, char *str, size_t sz) {
  */
 int H3_EXPORT(isValidCell)(H3Index h) {
-    if (H3_GET_HIGH_BIT(h) != 0) return 0;
-
-    if (H3_GET_MODE(h) != H3_HEXAGON_MODE) return 0;
-
-    if (H3_GET_RESERVED_BITS(h) != 0) return 0;
-
-    int baseCell = H3_GET_BASE_CELL(h);
-    if (baseCell < 0 || baseCell >= NUM_BASE_CELLS) {  // LCOV_EXCL_BR_LINE
-        // Base cells less than zero can not be represented in an index
-        return 0;
-    }
-
-    int res = H3_GET_RESOLUTION(h);
-    if (res < 0 || res > MAX_H3_RES) {  // LCOV_EXCL_BR_LINE
-        // Resolutions less than zero can not be represented in an index
-        return 0;
-    }
-
-    bool foundFirstNonZeroDigit = false;
-    for (int r = 1; r <= res; r++) {
-        Direction digit = H3_GET_INDEX_DIGIT(h, r);
-
-        if (!foundFirstNonZeroDigit && digit != CENTER_DIGIT) {
-            foundFirstNonZeroDigit = true;
-            if (_isBaseCellPentagon(baseCell) && digit == K_AXES_DIGIT) {
+    // Implementation strategy:
+    //
+    // Walk from high to low bits, checking validity of
+    // groups of bits as we go. After each check, shift bits off
+    // to the left, so that the next relevant group is at the
+    // highest bit location in the H3 Index, which we can
+    // easily read off with the `GT` macro. This strategy helps
+    // us avoid re-computing shifts and masks for each group.
+    //
+    // |   Region   | # bits |
+    // |------------|--------|
+    // | High       |      1 |
+    // | Mode       |      4 |
+    // | Reserved   |      3 |
+    // | Resolution |      4 |
+    // | Base Cell  |      7 |
+    // | Digit 1    |      3 |
+    // | Digit 2    |      3 |
+    // | ...        |    ... |
+    // | Digit 15   |      3 |
+    //
+    // Additionally, we try to group operations and void loops when possible.
+
+    // The 1 high bit should be 0b0
+    // The 4 mode bits should be 0b0001 (H3_CELL_MODE)
+    // The 3 reserved bits should be 0b000
+    // In total, the top 8 bits should be 0b00001000
+    if (GT(h, 8) != 0b00001000) return 0;
+    h <<= 8;
+
+    // Number of bits in each of the resolution, base cell, and digit regions.
+    const int nBitsRes = 4;
+    const int nBitsBaseCell = 7;
+    const int nBitsDigit = H3_PER_DIGIT_OFFSET;
+
+    // No need to check resolution; any 4-bit number is a valid resolution.
+    const int res = GT(h, nBitsRes);
+    h <<= nBitsRes;
+
+    // Check that base cell number is valid.
+    const int bc = GT(h, nBitsBaseCell);
+    if (bc >= NUM_BASE_CELLS) return 0;
+    h <<= nBitsBaseCell;
+
+    // Now check that each resolution digit is valid.
+    // Let `r` denote the resolution we're currently checking.
+    int r = 1;
+
+    // Pentagon cells start with a sequence of 0's (CENTER_DIGIT's).
+    // The first nonzero digit can't be a 1 (i.e., "deleted subsequence",
+    // PENTAGON_SKIPPED_DIGIT, or K_AXES_DIGIT).
+    // Test for pentagon base cell first to avoid this loop if possible.
+    if (isBaseCellPentagonArr[bc]) {
```

Comment:
Using this lookup table seems to be a bit faster than calling `_isBaseCellPentagon(bc)`.

Reply:
Alternative:

```c
int _isBaseCellPentagon(int baseCell) {
    switch(baseCell) {
        case 4:
        case 14:
        case 24:
        case 38:
        case 49:
        case 58:
        case 63:
        case 72:
        case 83:
        case 97:
        case 107:
        case 117:
            return 1;
        default:
            return 0;
    }
}
```

Reply:
If you can use `baseCellData[bc].isPentagon` that should be equivalent? Not sure of the linking cost of going outside the file.